torch/lib/THD/base/data_channels/Store.cpp

```diff
@@ -0,0 +1,205 @@
+#include "Store.hpp"
+#include "../ChannelUtils.hpp"
+
+#include <poll.h>
+#include <system_error>
+
+namespace thd {
+
+using size_type = uint64_t;
+using store_type = std::unordered_map<std::string, std::vector<char>>;
+
+static void oracle_daemon();
+static bool query(int socket);
+static bool check(const store_type& store,
+                  const std::vector<std::string>& keys);
+static void wakeup(int socket);
+
+
+enum query_type : uint8_t {
+  set,
+  get,
+  wait,
+  finish,
+  stop_waiting
+};
+ 
+Store::Store()
+    : rank(load_rank_env()) {
+  // Master runs the oracle
+  if (rank == 0) {
+    std::thread oracle(oracle_daemon);
+    oracle.detach();
+  }
+  // everyone introduces himself to the oracle
+  std::tie(master_addr, master_port) = load_worker_env();
+  socket = connect(master_addr, master_port);
+  send_bytes<rank_type>(socket, &rank, 1);
+}
+
+Store::~Store() {
+  // the 0 process finishes the daemon
+  query_type fin = query_type::finish;
+  if (rank == 0) 
+    send_bytes<query_type>(socket, &fin, 1);
+  // TODO unlucky interleave? Daemon might not finish
```

Comment:
To prevent that you can add `std::thread _oracle` variable to `Store` definition and create thread (in constructor) without detaching it:
```
std::thread tmp_thread(oracle_daemon);
std::swap(_oracle, tmp_thread);
```
This should work.

And then here just send the finish message and call `_oracle.join()`.